bulldozer/merge.go

```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
 		for i := 0; i < MaxPullRequestPollCount; i++ {
 			<-ticker.C
 
-			pr, _, err := client.PullRequests.Get(ctx, pullCtx.Owner(), pullCtx.Repo(), pullCtx.Number())
+			mergeState, err := pullCtx.MergeState(ctx)
 			if err != nil {
-				logger.Error().Err(errors.WithStack(err)).Msgf("Failed to retrieve pull request %q", pullCtx.Locator())
+				logger.Error().Err(err).Msgf("Failed to get merge state for %q", pullCtx.Locator())
 				return
 			}
 
-			if pr.GetState() == "closed" {
+			if mergeState.Closed {
 				logger.Debug().Msg("Pull request already closed")
 				return
 			}
 
-			if pr.Mergeable == nil {
+			if mergeState.Mergeable == nil {
 				logger.Debug().Msg("Pull request mergeability not yet known")
 				continue
 			}
 
-			if !pr.GetMergeable() {
+			if !*mergeState.Mergeable {
 				logger.Debug().Msg("Pull request is not mergeable")
 				return
 			}
 
 			// Try a merge, a 405 is expected if required reviews are not satisfied
 			logger.Info().Msgf("Attempting to merge pull request with method %s", mergeOpts.MergeMethod)
-			result, _, err := client.PullRequests.Merge(ctx, pullCtx.Owner(), pullCtx.Repo(), pullCtx.Number(), commitMessage, mergeOpts)
+			sha, err := merger.Merge(ctx, pullCtx, commitMessage, mergeOpts)
 			if err != nil {
-				gerr, ok := err.(*github.ErrorResponse)
+				gerr, ok := errors.Cause(err).(*github.ErrorResponse)
 				if !ok {
-					logger.Error().Err(errors.WithStack(err)).Msg("Merge failed unexpectedly")
+					logger.Error().Err(err).Msg("Merge failed unexpectedly")
 					continue
 				}
 
 				switch gerr.Response.StatusCode {
 				case http.StatusMethodNotAllowed:
-					logger.Info().Msgf("Merge rejected due to unsatisfied condition %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to unsatisfied condition: %q", gerr.Message)
 					return
 				case http.StatusConflict:
-					logger.Info().Msgf("Merge rejected due to being invalid %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to being invalid: %q", gerr.Message)
 					return
 				default:
-					logger.Error().Err(errors.WithStack(err)).Msgf("Merge failed unexpectedly %q", gerr.Message)
+					logger.Error().Msgf("Merge failed with unexpected status: %d: %q", gerr.Response.StatusCode, gerr.Message)
```

Comment:
losing the stacktrace on purpose?

Reply:
Yeah, we already know that the error in this case is an HTTP failure, so the important part is the status code and the message. It's also unambiguous where the error is generated.